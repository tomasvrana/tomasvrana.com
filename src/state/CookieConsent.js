import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'
import { getLocalStorageValue, setLocalStorageValue } from '../helpers/localStorage'
import { allowTracking } from '../helpers/analytics'

const CookieConsentContext = createContext()
const Consumer = CookieConsentContext.Consumer

const Provider = ({ children }) => {
  const [ allowCookies, setAllowCookies ] = useState(getLocalStorageValue('allowCookies') === 'true')

  const updateAllowCookies = allow => {
    if (allow) {
      allowTracking()
      setLocalStorageValue('allowCookies', 'true')
    }

    setAllowCookies(allow)
  }

  return (
    <CookieConsentContext.Provider value={{ allowCookies, setAllowCookies: updateAllowCookies }}>
      {children}
    </CookieConsentContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
