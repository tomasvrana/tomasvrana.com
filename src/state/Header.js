import React, { createContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { getURIPathWithoutLang } from '../helpers/url'

const HeaderContext = createContext()
const Consumer = HeaderContext.Consumer

const pageChangeListeners = []
const onPageChange = () => pageChangeListeners.forEach(listener => listener())

const availableKeys = [
  'byron',
  'shelley',
  'goguen',
  'updates'
]

const getHeaderContentKey = (lang) => {
  const uri = getURIPathWithoutLang(lang)
  const section = uri.replace(/^\//, '').split('/').shift()
  return availableKeys.includes(section) ? section : ''
}

const Provider = ({ children, lang }) => {
  const [ key, setKey ] = useState(getHeaderContentKey(lang))

  const onPageChange = () => setKey(getHeaderContentKey(lang))
  useEffect(() => {
    pageChangeListeners.push(onPageChange)

    return () => pageChangeListeners.pop()
  }, [])

  return (
    <HeaderContext.Provider value={{ key }}>
      {children}
    </HeaderContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any,
  lang: PropTypes.string.isRequired
}

export {
  Provider,
  Consumer,
  onPageChange
}
