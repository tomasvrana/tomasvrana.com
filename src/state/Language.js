import React, { createContext, useState, useEffect } from 'react'
import { navigate } from 'gatsby'
import PropTypes from 'prop-types'
import config from '../config'
import * as moment from 'moment'
import { getURIPath, getHash, languageSetInURL } from '../helpers/url'
import { getLocalStorageValue, setLocalStorageValue } from '../helpers/localStorage'
import { getNavigatorLanguage } from '../helpers/navigator'
import { autoCapture } from '../helpers/analytics'
import { LANGUAGE } from '../constants/analytics'

const LanguageContext = createContext()
const Consumer = LanguageContext.Consumer

function getUrlLang () {
  const [ urlLang ] = getURIPath().split('/').splice(1)
  return urlLang
}

function getDefaultLanguage () {
  // Use URL language where available
  if (config.availableLanguages[getUrlLang()]) return getUrlLang()
  // Fallback to local storage or default config
  const localStorageLang = getLocalStorageValue('lang')
  // Users browser language
  const navigatorLanguage = getNavigatorLanguage().substring(0, 2)
  // Use local storage first as they have already visited the site
  if (localStorageLang && config.availableLanguages[localStorageLang]) return localStorageLang
  // Use the users browser language if it supported by the site
  if (navigatorLanguage && config.availableLanguages[navigatorLanguage]) return navigatorLanguage
  // Fall back to the first available language in the config
  return Object.keys(config.availableLanguages)[0]
}

const Provider = ({ children }) => {
  const [ lang, updateLang ] = useState(getDefaultLanguage())
  const setLang = language => {
    if (!config.availableLanguages[language]) throw new Error(`Invalid language ${language}`)
    const pathParts = getURIPath().split('/').splice(1)
    if (languageSetInURL()) {
      pathParts.splice(0, 1, language)
    } else {
      pathParts.unshift(language)
    }

    setLocalStorageValue('lang', language)
    moment.locale(language)
    updateLang(language)
    if (lang !== language) autoCapture({ category: LANGUAGE, label: language, action: 'updated_language' })
    navigate(`/${pathParts.join('/')}${getHash()}`)
  }

  useEffect(() => {
    setLang(lang)
  }, [])

  return (
    <LanguageContext.Provider
      value={{
        lang,
        setLang,
        availableLocales: Object.keys(config.availableLanguages).map(key => config.availableLanguages[key].locale),
        locale: config.availableLanguages[lang].locale
      }}
    >
      {children}
    </LanguageContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
