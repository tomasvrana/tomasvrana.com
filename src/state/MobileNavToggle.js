import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'

const MobileNavToggleContext = createContext()
const Consumer = MobileNavToggleContext.Consumer

const Provider = ({ children }) => {
  const [ toggle, updateToggle ] = useState(false)

  const toggleMobileNav = () => {
    updateToggle(!toggle)
  }

  return (
    <MobileNavToggleContext.Provider value={{ toggle, toggleMobileNav }}>
      {children}
    </MobileNavToggleContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
