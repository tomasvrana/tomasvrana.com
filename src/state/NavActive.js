import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'

const NavActiveContext = createContext()
const Consumer = NavActiveContext.Consumer

const Provider = ({ children }) => {
  const [ navActive, updateNavActive ] = useState('')

  return (
    <NavActiveContext.Provider value={{ navActive, updateNavActive }}>
      {children}
    </NavActiveContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
