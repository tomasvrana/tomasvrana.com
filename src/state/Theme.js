import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'
import themes from '../config/themes'
import { getLocalStorageValue, setLocalStorageValue } from '../helpers/localStorage'

const ThemeContext = createContext()
const Consumer = ThemeContext.Consumer

function getInitialTheme () {
  const localStorageTheme = getLocalStorageValue('theme')
  return (localStorageTheme && themes[localStorageTheme])
    ? localStorageTheme
    : Object.keys(themes)[0]
}

const Provider = ({ children }) => {
  const [ theme, updateTheme ] = useState(getInitialTheme())

  const setTheme = theme => {
    if (!Object.keys(themes).includes(theme)) throw new Error(`Theme ${theme} does not exist`)
    setLocalStorageValue('theme', theme)
    updateTheme(theme)
  }

  return (
    <ThemeContext.Provider value={{ themeString: theme, theme: themes[theme], setTheme }}>
      {children}
    </ThemeContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
