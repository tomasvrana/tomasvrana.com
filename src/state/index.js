import {
  Consumer as CookieConsentConsumer,
  Provider as CookieConsentProvider
} from './CookieConsent'
import {
  Consumer as HeaderConsumer,
  Provider as HeaderProvider
} from './Header'
import {
  Consumer as LanguageConsumer,
  Provider as LanguageProvider
} from './Language'
import {
  Consumer as PageLoaderConsumer,
  Provider as PageLoaderProvider
} from './PageLoader'
import {
  Consumer as ThemeConsumer,
  Provider as ThemeProvider
} from './Theme'

export {
  CookieConsentConsumer,
  CookieConsentProvider,
  HeaderConsumer,
  HeaderProvider,
  LanguageConsumer,
  LanguageProvider,
  PageLoaderConsumer,
  PageLoaderProvider,
  ThemeConsumer,
  ThemeProvider
}
