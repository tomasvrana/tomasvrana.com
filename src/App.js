import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider, ThemeConsumer, LanguageProvider, LanguageConsumer, PageLoaderProvider, CookieConsentProvider, HeaderProvider } from './state'
import { ThemeProvider as StyledThemeProvider } from 'styled-components'
import PickerContainer from './components/PickerContainer'
import Header from './components/Header'
import GlobalStyles from './GlobalStyles'

const App = ({ element }) => (
  <CookieConsentProvider>
    <ThemeProvider>
      <ThemeConsumer>
        {({ theme }) => (
          <StyledThemeProvider theme={theme}>
            <Fragment>
              <GlobalStyles />
              <LanguageProvider>
                <PageLoaderProvider>
                  <Fragment>
                    <PickerContainer />
                    <LanguageConsumer>
                      {({ lang }) => (
                        <HeaderProvider lang={lang}>
                          <Header />
                        </HeaderProvider>
                      )}
                    </LanguageConsumer>
                    {element}
                  </Fragment>
                </PageLoaderProvider>
              </LanguageProvider>
            </Fragment>
          </StyledThemeProvider>
        )}
      </ThemeConsumer>
    </ThemeProvider>
  </CookieConsentProvider>
)

App.propTypes = {
  element: PropTypes.node.isRequired
}

export default App
