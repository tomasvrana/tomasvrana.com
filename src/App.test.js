import React from 'react'
import { shallow } from 'enzyme'
import App from './App'
import * as state from './state'

const getWrapper = props => {
  return shallow(
    <App
      element={(<p>Element</p>)}
      {...props}
    />
  ).dive()
}

describe('App', () => {
  let context
  beforeEach(() => {
    context = { theme: { example: 'theme' } }
    state.ThemeConsumer = jest.fn(props => props.children(context))
  })

  test('it renders correctly', () => {
    const wrapper = getWrapper()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.children().children().dive()).toMatchSnapshot()
  })
})
