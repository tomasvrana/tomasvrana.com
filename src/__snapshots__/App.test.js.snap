// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App it renders correctly 1`] = `
<ContextProvider
  value={
    Object {
      "allowCookies": false,
      "setAllowCookies": [Function],
    }
  }
>
  <Provider>
    <mockConstructor>
      <Component />
    </mockConstructor>
  </Provider>
</ContextProvider>
`;

exports[`App it renders correctly 2`] = `
<ThemeProvider
  theme={
    Object {
      "example": "theme",
    }
  }
>
  <GlobalStyleComponent />
  <Provider>
    <Provider>
      <_default />
      <PickerContainer />
      <_default />
      <p>
        Element
      </p>
      <Footer />
    </Provider>
  </Provider>
</ThemeProvider>
`;
